File: sdks/php/src/Api/Segments.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\CreateSegmentBody;
use Apollo\Deploy\Signal\Sdk\CursorPage;
use Apollo\Deploy\Signal\Sdk\SegmentPageResponse;
use Apollo\Deploy\Signal\Sdk\SegmentResponse;
use Psr\Http\Message\ResponseInterface;

/** SegmentsAPI domain operations. */
class Segments
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $projectId
     * @return SegmentPageResponse
     * @throws SDKError
     */
    public function listSegments(
        string $projectId,
    ): SegmentPageResponse
    {
        $path = '/v1/projects/${projectId}/segments';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(SegmentPageResponse::class, 'fromArray')) {
            return SegmentPageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $segmentId
     * @return SegmentResponse
     * @throws SDKError
     */
    public function getSegment(
        string $projectId,
        string $segmentId,
    ): SegmentResponse
    {
        $path = '/v1/projects/${projectId}/segments/${segmentId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${segmentId}', rawurlencode(Transport::wireValue($segmentId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(SegmentResponse::class, 'fromArray')) {
            return SegmentResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $segmentId
     * @return CursorPage
     * @throws SDKError
     */
    public function listContactsInSegment(
        string $projectId,
        string $segmentId,
    ): CursorPage
    {
        $path = '/v1/projects/${projectId}/segments/${segmentId}/contacts';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${segmentId}', rawurlencode(Transport::wireValue($segmentId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(CursorPage::class, 'fromArray')) {
            return CursorPage::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param CreateSegmentBody $input
     * @return SegmentResponse
     * @throws SDKError
     */
    public function createSegment(
        string $projectId,
        CreateSegmentBody $input,
    ): SegmentResponse
    {
        $path = '/v1/projects/${projectId}/segments';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(SegmentResponse::class, 'fromArray')) {
            return SegmentResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $segmentId
     * @return void
     * @throws SDKError
     */
    public function deleteSegment(
        string $projectId,
        string $segmentId,
    ): void
    {
        $path = '/v1/projects/${projectId}/segments/${segmentId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${segmentId}', rawurlencode(Transport::wireValue($segmentId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/segments.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ CreateSegmentBody, CursorPage, SegmentPageResponse, SegmentResponse };

/// SegmentsAPI API operations.
pub struct SegmentsAPI {
    transport: Transport,
}

impl SegmentsAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn list_segments(
        &self,
        project_id: String,
    ) -> Result<SegmentPageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/segments".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<SegmentPageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_segment(
        &self,
        project_id: String,
        segment_id: String,
    ) -> Result<SegmentResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/segments/${segmentId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${segmentId}", &encode_path(&segment_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<SegmentResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn list_contacts_in_segment(
        &self,
        project_id: String,
        segment_id: String,
    ) -> Result<CursorPage, SdkError> {
        let mut path = "/v1/projects/${projectId}/segments/${segmentId}/contacts".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${segmentId}", &encode_path(&segment_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<CursorPage>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn create_segment(
        &self,
        project_id: String,
        body: &CreateSegmentBody,
    ) -> Result<SegmentResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/segments".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<SegmentResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn delete_segment(
        &self,
        project_id: String,
        segment_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/segments/${segmentId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${segmentId}", &encode_path(&segment_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
